atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/qos/ThriftQueryWeighers.java

```diff
@@ -0,0 +1,196 @@
+                        .from(DEFAULT_ESTIMATED_WEIGHT)
+                        .numBytes(ESTIMATED_NUM_BYTES_PER_ROW * numberOfQueriedRows)
+                        .timeTakenNanos(timeTakenNanos)
+                        .build();
+            }
+        };
+    }
+
+    private static <T> QosClient.QueryWeigher<T> readWeigher(Function<T, Long> bytesRead,
+            Function<T, Integer> numRows,
+            int numberOfQueriedRows) {
+        return new QosClient.QueryWeigher<T>() {
+            @Override
+            public QueryWeight estimate() {
+                return ImmutableQueryWeight.builder()
+                        .from(DEFAULT_ESTIMATED_WEIGHT)
+                        .numBytes(ESTIMATED_NUM_BYTES_PER_ROW * numberOfQueriedRows)
+                        .build();
+            }
+
+            @Override
+            public QueryWeight weighSuccess(T result, long timeTakenNanos) {
+                return ImmutableQueryWeight.builder()
+                        .numBytes(safeGetNumBytesOrDefault(() -> bytesRead.apply(result)))
+                        .timeTakenNanos(timeTakenNanos)
+                        .numDistinctRows(numRows.apply(result))
+                        .build();
+            }
+
+            @Override
+            public QueryWeight weighFailure(Exception error, long timeTakenNanos) {
+                return ImmutableQueryWeight.builder()
+                        .from(estimate())
+                        .timeTakenNanos(timeTakenNanos)
+                        .build();
+            }
+        };
+    }
+
+    private static <T> QosClient.QueryWeigher<T> writeWeigher(long numRows, Supplier<Long> bytesWritten) {
+        Supplier<Long> weight = Suppliers.memoize(() -> safeGetNumBytesOrDefault(bytesWritten))::get;
+
+        return new QosClient.QueryWeigher<T>() {
+            @Override
+            public QueryWeight estimate() {
+                return ImmutableQueryWeight.builder()
+                        .from(DEFAULT_ESTIMATED_WEIGHT)
+                        .numBytes(weight.get())
+                        .numDistinctRows(numRows)
+                        .build();
+            }
+
+            @Override
+            public QueryWeight weighSuccess(T result, long timeTakenNanos) {
+                return ImmutableQueryWeight.builder()
+                        .from(estimate())
+                        .timeTakenNanos(timeTakenNanos)
+                        .build();
+            }
+
+            @Override
+            public QueryWeight weighFailure(Exception error, long timeTakenNanos) {
+                return ImmutableQueryWeight.builder()
+                        .from(estimate())
+                        .timeTakenNanos(timeTakenNanos)
+                        .build();
+            }
+        };
+    }
+
+    // TODO(nziebart): we really shouldn't be needing to catch exceptions here
+    private static long safeGetNumBytesOrDefault(Supplier<Long> numBytes) {
```

Comment:
Most kinds of thrift objects with different fields filled in or empty have been tested, but I wouldnt stay the tests are comprehensive and cover all kinds of objects. Do we want to add more tests, is there some way to just generate a bunch of objects and ensure that the size calculating method doesn't throw?

Reply:
i would think that the cassandra ETE tests would catch these errors. but i'm ok with leaving this for now, and removing it fairly soon after we can verify that this error doesn't occur in the field.